pkg/conversion/conversion_rules.go

```diff
@@ -0,0 +1,208 @@
+	}
+
+	if rule.Transformation == "" {
+		return cr, nil
+	}
+
+	env, err := getCELEnv(schemaForField)
+	if env == nil {
+		err = errors.New("environment was nil")
+	}
+	if err != nil {
+		return nil, field.InternalError(rulePath.Child("field"), fmt.Errorf("error creating CEL environment: %w", err))
+	}
+
+	ast, issues := env.Compile(rule.Transformation)
+	if issues != nil {
+		return nil, field.Invalid(rulePath.Child("transformation"), rule.Transformation, fmt.Sprintf("error compiling CEL program: %v", issues.Err()))
+	}
+
+	program, err := env.Program(ast,
+		cel.EvalOptions(cel.OptOptimize),
+		cel.OptimizeRegex(library.ExtensionLibRegexOptimizations...),
+		cel.InterruptCheckFrequency(celCheckFrequency),
+	)
+	if err != nil {
+		return nil, field.Invalid(rulePath.Child("transformation"), rule.Transformation, fmt.Sprintf("error creating CEL program: %v", err))
+	}
+
+	cr.Program = program
+
+	return cr, nil
+}
+
+func getCELEnv(structuralSchema *structuralschema.Structural) (*cel.Env, error) {
+	env, err := cel.NewEnv(cel.HomogeneousAggregateLiterals())
+	if err != nil {
+		return nil, fmt.Errorf("error creating CEL environment: %w", err)
+	}
+
+	registry := model.NewRegistry(env)
+
+	// inline local copy of upstream's generateUniqueSelfTypeName()
+	scopedTypeName := fmt.Sprintf("selfType%d", time.Now().Nanosecond())
+
+	ruleTypes, err := model.NewRuleTypes(scopedTypeName, structuralSchema, true, registry)
+	if err != nil {
+		return nil, fmt.Errorf("error creating rule types: %w", err)
+	}
+	if ruleTypes == nil {
+		return nil, nil
+	}
+
+	opts, err := ruleTypes.EnvOptions(env.TypeProvider())
+	if err != nil {
+		return nil, fmt.Errorf("error getting CEL environment options: %w", err)
+	}
+
+	root, ok := ruleTypes.FindDeclType(scopedTypeName)
+	if !ok {
+		rootDecl := model.SchemaDeclType(structuralSchema, true)
+		if rootDecl == nil {
+			return nil, fmt.Errorf("unable to find CEL decl type for %s", structuralSchema.Type)
+		}
+		root = rootDecl.MaybeAssignTypeName(scopedTypeName)
+	}
+
+	var propDecls []*expr.Decl
+	propDecls = append(propDecls, decls.NewVar("self", root.ExprType()))
+
+	opts = append(opts, cel.Declarations(propDecls...), cel.HomogeneousAggregateLiterals())
+	opts = append(opts, library.ExtensionLibs...)
+	return env.Extend(opts...)
+}
+
+func schemaIndexByJSONPath(s *structuralschema.Structural, path string) (*structuralschema.Structural, error) {
+	cursor := s
+
+	fields := strings.Split(path, ".")
+	visited := make([]string, 0, len(fields))
+
+	// spec.name.first ==> 1
+	//  0    1     2
+	lastParentIndex := len(fields) - 2
+
+	for i, f := range fields {
+		// While iterating, i-1 is the index of the parent field
+		if lastParentIndex >= 0 && (i-1) <= lastParentIndex && cursor.Type != "object" {
+			return nil, fmt.Errorf("expected field %q to be an object", strings.Join(visited, "."))
+		}
+
+		visited = append(visited, f)
+
+		property, exists := cursor.Properties[f]
```

Comment:
also `AdditionalProperties` is possible.

Reply:
Fixing